File: crates/types/src/settings/music.rs
use serde::{Deserialize, Serialize};

#[cfg(feature = "ts-rs")]
use ts_rs::TS;

// Frontend-facing typed view for the "music" settings domain.
// This defines player/platform selection and audio-related preferences
// to be stored under prefs.music.* via the settings service.

/// Platform selection mode for music source routing.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub enum MusicSourceMode {
    All,
    Single,
    Many,
}

impl Default for MusicSourceMode {
    fn default() -> Self { Self::All }
}

/// Selected platform(s) used by the app when searching/playing.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub struct MusicSourceSelection {
    pub mode: MusicSourceMode,
    /// Plugin IDs when mode is Single/Many. Empty for All.
    #[serde(default)]
    pub ids: Vec<String>,
}

impl Default for MusicSourceSelection {
    fn default() -> Self {
        Self { mode: MusicSourceMode::All, ids: Vec::new() }
    }
}

/// Playback related preferences (kept minimal; extend as needed).
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub struct MusicPlaybackSettings {
    /// Enable loudness normalization if supported by source.
    pub normalize: Option<bool>,
    /// Crossfade duration in milliseconds.
    pub crossfade_ms: Option<u32>,
    /// Prefer seamless (gapless) playback when possible.
    pub gapless: Option<bool>,
}

/// A single audio effect unit in the processing chain.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub struct MusicEffectUnit {
    pub id: String,
    #[serde(rename = "type")]
    pub type_: String,
    pub enabled: Option<bool>,
    /// Effect parameters (implementation-defined).
    #[cfg_attr(feature = "ts-rs", ts(type = "Record<string, any>"))]
    pub params: Option<serde_json::Value>,
}

/// Audio effects chain configuration.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub struct MusicEffectsSettings {
    pub enabled: Option<bool>,
    #[serde(default)]
    pub chain: Vec<MusicEffectUnit>,
}

/// Root of the "music" settings domain.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
#[cfg_attr(
    feature = "ts-rs",
    derive(TS),
    ts(export, export_to = "bindings.d.ts", rename_all = "camelCase")
)]
pub struct MusicSettings {
    /// Current platform selection (Home/all or specified plugins).
    pub source: Option<MusicSourceSelection>,
    /// Preferred order for non-Home sources (plugin IDs).
    pub sources_order: Option<Vec<String>>,
    /// Playback preferences.
    pub playback: Option<MusicPlaybackSettings>,
    /// Effects chain configuration.
    pub effects: Option<MusicEffectsSettings>,
}

File: src/atoms/settings/music.ts
import { createSettingAtom } from "~/atoms/helper/setting"
import type { DomainBinding } from "~/services/settings"
import { createDomainBinding } from "~/services/settings"
import type { MusicSettings } from "~/types/bindings"

type MusicDefaults = Strictify<MusicSettings>

export const createDefaultMusicSettings = (): MusicDefaults => ({
  // Platform/source selection
  source: {
    mode: "all",
    ids: [],
  },
  // User-preferred order for non-Home sources (plugin IDs)
  sourcesOrder: [],
  // Playback preferences
  playback: {
    normalize: false,
    crossfadeMs: 0,
    gapless: true,
  },
  // Audio effects chain configuration
  effects: {
    enabled: false,
    chain: [],
  },
})

const {
  useSettingKey: useMusicSettingKey,
  useSettingSelector: useMusicSettingSelector,
  useSettingKeys: useMusicSettingKeys,
  setSetting: setMusicSetting,
  clearSettings: clearMusicSettings,
  initializeDefaultSettings: initializeDefaultMusicSettings,
  getSettings: getMusicSettings,
  useSettingValue: useMusicSettingValue,
  settingAtom: __musicSettingAtom,
} = createSettingAtom("music", createDefaultMusicSettings)

export {
  __musicSettingAtom,
  useMusicSettingKey,
  useMusicSettingSelector,
  useMusicSettingKeys,
  setMusicSetting,
  clearMusicSettings,
  initializeDefaultMusicSettings,
  getMusicSettings,
  useMusicSettingValue,
}

let __musicBinding: DomainBinding<MusicDefaults> | null = null

export function getMusicBinding() {
  if (!__musicBinding) {
    __musicBinding = createDomainBinding<MusicDefaults, MusicSettings>({
      domain: "music",
      defaultFactory: createDefaultMusicSettings,
      setLocal: (k, v) => setMusicSetting(k as any, v as any),
      keyMap: {} as any,
    })
  }
  return __musicBinding
}

export async function hydrateMusic() {
  await getMusicBinding().hydrate()
}

export function listenMusic(onAfterChange?: (k: keyof MusicDefaults, v: any) => void) {
  return getMusicBinding().listen(onAfterChange as any)
}

export function setMusic<K extends keyof MusicDefaults>(key: K, value: MusicDefaults[K]) {
  return getMusicBinding().set(key, value)
}

